networkx/algorithms/planarity.py

```diff
@@ -909,15 +908,86 @@ def neighbors_cw_order(self, v):
         node
 
         """
-        if len(self[v]) == 0:
+        if not self._succ[v]:
             # v has no neighbors
             return
-        start_node = self.nodes[v]["first_nbr"]
+        start_node = next(reversed(self._succ[v]))
         yield start_node
-        current_node = self[v][start_node]["cw"]
+        current_node = self._succ[v][start_node]["cw"]
         while start_node != current_node:
             yield current_node
-            current_node = self[v][current_node]["cw"]
+            current_node = self._succ[v][current_node]["cw"]
+
+    def add_half_edge(self, start_node, end_node, cw=None, ccw=None):
+        """Adds a half-edge from start_node to end_node.
+
+        If the half-edge is not the first one out of start_node, a reference
+        node must be provided either in the clockwise (parameter cw) or in
+        the counterclockwise (parameter ccw) direction. Only one of cw/cww
+        parameter can be specified (or neither in the case of the first edge).
+        Note that specifying a reference in the clockwise (cw) direction means
+        inserting the new edge in the first counterclockwise position with
+        respect to the reference (and vice-versa).
+
+        Parameters
+        ----------
+        start_node : node
+            Start node of inserted edge.
+        end_node : node
+            End node of inserted edge.
+        cw/ccw: node
+            End node of reference edge.
+
+        Raises
+        ------
+        NetworkXException
+            If the reference_neighbor does not exist.
+
+        See Also
+        --------
+        connect_components
+        """
+
+        if start_node in self._succ and self._succ[start_node]:
```

Comment:
If you are going to use `self._succ` for speed, you should probably make it local too:
```python
        succs = self._succ.get(start_node)  # handle start_node not in self._succ
        if succs:
```
with lots of changes later for `self._succ[start_node]` -> `succs`.  The name `succs` could be replaced by `nbrs` or something else more appropriate if you like. But it is used elsewhere in NetworkX IIRC. It's supposed to be an abbreviation of successors.  Feel free to choose something appropriate.

Reply:
done